README.md

```diff
@@ -1,3 +1,51 @@
 # kecho
 
 This is a lightweight echo server implementation in Linux kernel mode.
+
+## Usage
+
+Run make to build kecho and the benchmarking tool:
+```
+$ make
+```
+
+After the module is loaded, you can use `telnet` to interact with the module:
+```
+$ telnet 127.0.0.1 12345
+```
+
+If the default port (`12345`) is in use or you simply want to use another port, you can specify it when loading the module. Also, `backlog` is specifiable, which has default value `128`:
+```
+$ sudo insmod fastecho.ko port={port_you_want} backlog={amount_you_want}
+```
+
+Also, you can start the benchmarking by running:
+```
+$ ./bench
+```
+
+Note that too much concurrent connection would be treat as sort of DDOS attack (mainly caused by configs (kernel: `tcp_max_syn_backlog` and `somaxconn`. Application (fastecho): `backlog`)). Nevertheless, default maximum number of fd per-process is `1024`. These limitation can cause performance degration when benchmarking, if you insist to proceed the benchmark, perform following modifications:
+
+- Use following commands to adjust kernel attributes:
+    ```
+    $ sudo sysctl net.core.somaxconn={depends_on_MAX_THREAD}
+    ```
+    ```
+    $ sudo sysctl net.ipv4.tcp_max_syn_backlog={ditto}
+    ```
+    Note that `$ sudo sysctl net.core.somaxconn` can get current value. `somaxconn` is max amount of established connection, whereas `tcp_max_syn_backlog` is max amount of connection at first step of TCP 3-way handshake (SYN).
+
+- Use following command to enlarge limitation of fd per-process:
+  ```
+  $ ulimit -n {ditto}
```

Comment:
Why `ditto` here?

Reply:
The maximum number of fd per-process can open is also a limitation to the benchmarking, since if we have `MAX_THREAD` larger than default `ulimit -n` (1024 by default), some benchmarking workers may get stalled on creation of the socket.

Reply:
Sorry, I'm wrong, `socket()` should return `EMFILE` instead of stalled.